File: DLL/Message.cpp
#include "pch.h"
#include "Message.h"


int Message::clientID = 0;


void Message::send(tcp::socket& s, int to, int from, int type, const wstring& data)
{
	Message m(to, from, type, data);
	m.send(s);
}


Message Message::send(int to, int type, const wstring& data)
{
	boost::asio::io_context io;
	tcp::socket s(io);
	tcp::resolver r(io);
	boost::asio::connect(s, r.resolve("127.0.0.1", "12345"));

	Message m(to, clientID, type, data);
	m.send(s);
	m.receive(s);
	if (type == MT_INIT) {
		clientID = m.header.to;
	}

	return m;
}

File: PythonClient/Message.py
import threading
from dataclasses import dataclass
import socket, struct, time, sys

MT_INIT = 0
MT_EXIT = 1
MT_GETDATA = 2
MT_DATA = 3
MT_NODATA = 4
MT_CONFIRM = 5

MR_BROKER = 10
MR_ALL = 50
MR_USER = 100


@dataclass
class MsgHeader:
    To: int = 0
    From: int = 0
    Type: int = 0
    Size: int = 0

    def Send(self, s):
        s.send(struct.pack(f"iiii", self.To, self.From, self.Type, self.Size))

    def Receive(self, s):
        try:
            (self.To, self.From, self.Type, self.Size) = struct.unpack(
                "iiii", s.recv(16)
            )
        except:
            self.Size = 0
            self.Type = MT_NODATA


class Message:
    ClientID = 0

    def __init__(self, To=0, From=0, Type=MT_DATA, Data=""):
        self.Header = MsgHeader(To, From, Type, len(Data) * 2)
        self.Data = Data

    def Send(self, s):
        self.Header.Send(s)
        if self.Header.Size > 0:
            s.send(struct.pack(f"{self.Header.Size}s", self.Data.encode("utf-16-le")))

    def Receive(self, s):
        self.Header.Receive(s)
        if self.Header.Size > 0:
            self.Data = struct.unpack(f"{self.Header.Size}s", s.recv(self.Header.Size))[
                0
            ].decode("utf-16-le")

    @staticmethod
    def SendMessage(To, Type=MT_DATA, Data=""):
        if len(sys.argv) > 1:
            HOST = sys.argv[1]
        else:
            HOST = "localhost"
        PORT = 12345
        with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:
            s.connect((HOST, PORT))
            m = Message(To, Message.ClientID, Type, Data)
            m.Send(s)
            m.Receive(s)
            if Type == MT_INIT:
                Message.ClientID = m.Header.To
            return m
